pylint/checkers/classes/special_methods_checker.py

```diff
@@ -16,11 +20,16 @@
     only_required_for_messages,
     safe_infer,
 )
+from pylint.lint.pylinter import PyLinter
 
 NEXT_METHOD = "__next__"
 
 
-def _safe_infer_call_result(node, caller, context=None):
+def _safe_infer_call_result(
+    node: nodes.AsyncFunctionDef | nodes.FunctionDef,
+    caller: nodes.AsyncFunctionDef | nodes.FunctionDef,
+    context: Any | None = None,
+) -> Any:
```

Comment:
This should be some form of `InferenceResult`.

Reply:
``-> Iterator[nodes.NodeNG | Uninferable | None] | None:`` seems to work, the typing in astroid is currently:

```
    def infer_call_result(self, caller=None, context=None):
        """Infer what the function returns when called.

        :returns: What the function returns.
        :rtype: iterable(NodeNG or Uninferable) or None
        """
```

Reply:
That's incorrect (_Never trust an astroid docstring 😄_).

There is at least one call to `infer()` in `infer_call_result` without any assertions so that can yield an `InferenceResult`.
I think we should leave this signature until we have `infer_call_result` typed correctly.